crates/subspace-farmer/src/single_disk_plot/piece_reader.rs

```diff
@@ -0,0 +1,102 @@
+use bitvec::prelude::*;
+use futures::channel::{mpsc, oneshot};
+use futures::SinkExt;
+use std::num::{NonZeroU16, NonZeroU32};
+use subspace_core_primitives::{Piece, PublicKey, SectorId, SectorIndex, PIECE_SIZE};
+use subspace_solving::derive_chunk_otp;
+
+#[derive(Debug)]
+pub(super) struct ReadPieceRequest {
+    pub(super) sector_index: SectorIndex,
+    pub(super) piece_offset: u64,
+    pub(super) response_sender: oneshot::Sender<Option<Piece>>,
+}
+
+/// Wrapper data structure that can be used to read pieces from single disk plot
+#[derive(Debug, Clone)]
+pub struct PieceReader {
+    read_piece_sender: mpsc::Sender<ReadPieceRequest>,
+}
+
+impl PieceReader {
+    pub(super) fn new() -> (Self, mpsc::Receiver<ReadPieceRequest>) {
+        let (read_piece_sender, read_piece_receiver) = mpsc::channel::<ReadPieceRequest>(10);
+
+        (Self { read_piece_sender }, read_piece_receiver)
+    }
+
+    pub(super) fn close_all_readers(&mut self) {
+        self.read_piece_sender.close_channel();
+    }
+
+    /// Read piece from sector by offset, `None` means input parameters are incorrect or piece
+    /// reader was shut down
+    pub async fn read_piece(
+        &mut self,
+        sector_index: SectorIndex,
+        piece_offset: u64,
+    ) -> Option<Piece> {
+        let (response_sender, response_receiver) = oneshot::channel();
+        self.read_piece_sender
+            .send(ReadPieceRequest {
+                sector_index,
+                piece_offset,
+                response_sender,
+            })
+            .await
+            .ok()?;
+        response_receiver.await.ok()?
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
+pub(super) fn read_piece(
+    sector_index: SectorIndex,
+    piece_offset: u64,
+    sector_count: u64,
+    public_key: &PublicKey,
+    first_sector_index: SectorIndex,
+    plot_sector_size: u64,
+    record_size: NonZeroU32,
+    space_l: NonZeroU16,
+    global_plot: &[u8],
+) -> Option<Piece> {
+    if sector_index < first_sector_index {
+        return None;
```

Comment:
please, add actual log messages. I couldn't receive pieces because of one of the `return None` conditions but I had to add them myself.